### PR TITLE
[Mono] Fix memory leak during data registration

### LINKED ISSUE #21035: Memory leak in mini_get_rgctx_entry_slot -> lookup_or_register
## Steps to Reproduce

1. https://github.com/mono/mono/blob/bdd772531d379b4e78593587d15113c37edd4a64/mono/mini/mini-generic-sharing.c#L2843
2. if the data is found in the lookup aka it is registered, the does nothing to ensure it frees `data`.
3. `data` is always newly allocated before passed to the method

### Current Behavior

Leak

### Expected Behavior

No leak

## On which platforms did you notice this

[*] macOS
[ ] Linux
[ ] Windows

**Version Used**:

Mono JIT compiler version 6.12.0.125 (2020-02/8c552e98bd6 Mon Mar  8 13:06:52 EST 2021)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	Interpreter:   yes
	LLVM:          yes(610)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)

## Stacktrace

```
Please paste the stack trace here if available.
```

image

### PR DESCRIPTION
**Pull request by fanyang-mono**
Fixes https://github.com/mono/mono/issues/21035. After this change is merged, it needs to be ported to https://github.com/mono/mono.

### CODE REVIEW

File: src/mono/mono/mini/mini-generic-sharing.c [hunk 1]
```diff
@@ -4302,9 +4307,32 @@ mini_get_rgctx_entry_slot (MonoJumpInfoRgctxEntry *entry)
...
+		case MONO_PATCH_INFO_VIRT_METHOD:
+			g_free ((MonoJumpInfoVirtMethod *) entry_data);
+			break;
+		case MONO_PATCH_INFO_DELEGATE_TRAMPOLINE:
+			g_free ((MonoDelegateClassMethodPair *) entry_data);
+			break;
+		case MONO_PATCH_INFO_GSHAREDVT_METHOD: {
+			for (int i = 0; i < ((MonoGSharedVtMethodInfo *) entry_data)->num_entries; ++i) {
+				g_free (&((MonoGSharedVtMethodInfo *) entry_data)->entries[i]);
+			}
```
Comment (lambdageek) [inline src/mono/mono/mini/mini-generic-sharing.c:4328]:
```suggestion
			g_free (((MonoGSharedVtMethodInfo *) entry_data)->entries);
```

It's one malloc - so there should be one free.

File: src/mono/mono/mini/mini-generic-sharing.c [hunk 2]
```diff
@@ -4302,9 +4307,32 @@ mini_get_rgctx_entry_slot (MonoJumpInfoRgctxEntry *entry)
 	MonoJitMemoryManager *jit_mm = get_default_jit_mm ();
 
 	if (entry->in_mrgctx)
-		return lookup_or_register_info (jit_mm->mem_manager, entry->d.method->klass, entry->d.method, entry->in_mrgctx, entry_data, entry->info_type, mono_method_get_context (entry->d.method));
+		return lookup_or_register_info (jit_mm->mem_manager, entry->d.method->klass, entry->d.method, entry->in_mrgctx, entry_data, &did_register, entry->info_type, mono_method_get_context (entry->d.method));
 	else
-		return lookup_or_register_info (jit_mm->mem_manager, entry->d.klass, NULL, entry->in_mrgctx, entry_data, entry->info_type, mono_class_get_context (entry->d.klass));
+		return lookup_or_register_info (jit_mm->mem_manager, entry->d.klass, NULL, entry->in_mrgctx, entry_data, &did_register, entry->info_type, mono_class_get_context (entry->d.klass));
```
Comment (lambdageek) [inline src/mono/mono/mini/mini-generic-sharing.c:4312]: Both of these branches `return`, so your cleanup code never gets called!

**Review by lambdageek — CHANGES_REQUESTED** [review]
Generally this is the right approach, but there's a couple of issues

File: src/mono/mono/mini/mini-generic-sharing.c (hunk 2 above)
Comment (fanyang-mono) [reply src/mono/mono/mini/mini-generic-sharing.c:4312]: Oops! Let me fix this.

**Review by lambdageek — APPROVED** [review]

**Pull request merged**